### ISSUE TITLE
Making transpose keybind work on selections

### ISSUE THREAD

**Issue by aadibajpai** [zed-industries/zed#5315]
The transpose keybind (ctrl+t) isn't really super useful right now but a feature I have oft wished for is the ability to swap pieces of code since it is very clumsy in most editors: you copy one part and paste it right next to the other then cut that other part or something of that order. 

It would be great if transpose worked on selections pretty much as a way to swap them.

**Comment by scorphus** [comment]
Any updates on this? This is a much needed feature and I'm willing to contribute.

But I need help.

- Is it possible to adapt the [`transpose`](https://github.com/zed-industries/zed/blob/675673ed5462711916e7824528698fc235d679b2/crates/editor/src/editor.rs#L6737) func?
- Is it possible to create such extension?
- Are these the only two options?
- Or what would be the best approach?

Thanks!

**Comment by notpeter** [comment]
I don't think it is currently practical to add this functionality given current [Extension Capabilities](https://zed.dev/docs/extensions/developing-extensions) and so this would likely have to be supported natively.

Given that the current functionality only swaps two adjacent character, is your use case that if a user has two selections the transpose function would just swap their contents?

**Comment by scorphus** [comment]
Yes. But not only two! It's also (very) useful when the user has multiple selections and wants to "rotate" them.

Take, for instance, a list of parameterized tests:

```python
@pytest.mark.parametrize(
    "foo, bar, expected",
    [
        ("foo string", 1, "xyz")
        ("another foo", 2, "123")
        ("some other foo", 3, "456")
        ("another foo too", 4, "abc")
        ("yet another foo", 5, "def")
        ("and another foo", 6, "ghi")
    ],
) 
```

If the user wants to add a new test between `bar` 3 and 4, they could simply add that line with `bar` being 7:

```python
@pytest.mark.parametrize(
    "foo, bar, expected",
    [
        ("foo string", 1, "xyz")
        ("another foo", 2, "123")
        ("some other foo", 3, "456")
        ("new foo", 7, "new")
        ("another foo too", 4, "abc")
        ("yet another foo", 5, "def")
        ("and another foo", 6, "ghi")
    ],
) 
```

Then they use multi-selection to select all the `bar` values from 7 to 6 (7, 4, 5 and 6) and rotate them three times (aka until `bar` values are ordered top down):

```python
@pytest.mark.parametrize(
    "foo, bar, expected",
    [
        ("foo string", 1, "xyz")
        ("another foo", 2, "123")
        ("some other foo", 3, "456")
        ("new foo", 4, "new")
        ("another foo too", 5, "abc")
        ("yet another foo", 6, "def")
        ("and another foo", 7, "ghi")
    ],
) 
```

This is just an illustration of a real use case I can remember right now. But there are times when such a feature saves you a whole lot of time.

**Comment by scorphus** [comment]
I forgot to mention that rotation in both directions, forward and backward, is a must have.

So, I'm eager to dip my fingers into Zed's crystal water! Maybe I can make it part of native features. Do you think you can give me some tips on where to start?

**Comment by notpeter** [comment]
This is absolutely not my area of expertise, but transpose is defined in crates/editor/src/editor.rs:
https://github.com/zed-industries/zed/blob/a0d4c82c0c155619f260d8f4e0c6b605a39ec8ae/crates/editor/src/editor.rs#L6804

And there are some tests in [crates/editor/src/editor_tests.rs](https://github.com/zed-industries/zed/blob/main/crates/editor/src/editor_tests.rs) with small well-defined operations.  For example here's a test for sort lines:
https://github.com/zed-industries/zed/blob/a0d4c82c0c155619f260d8f4e0c6b605a39ec8ae/crates/editor/src/editor_tests.rs#L3455-L3475

Your example is about rotating a single selection of multiple lines, I would start with that and hold off handling multiple selections.

**Comment by Esteban-Rocha** [comment]
Any updates? Transpose is like a very basic functionality, tying out Zed but this is a deal breaker for me.

**Comment by mattbaker** [comment]
@Esteban-Rocha based on the thread it sounds like the team is open to a PR and notpeter provided some initial code to explore. I don’t agree this is “very basic functionality” personally, but I can see it’s important to you, so it’s a great opportunity to contribute!

**Comment by Esteban-Rocha** [comment]
> [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/5914592?u=b90e71f1ace8010c6c16ac310fc5fb774df8b91c&amp;v=4&amp;size=80">@Esteban-Rocha](https://github.com/Esteban-Rocha?rgh-link-date=2025-02-14T17%3A00%3A02.000Z) based on the thread it sounds like the team is open to a PR and notpeter provided some initial code to explore. I don’t agree this is “very basic functionality” personally, but I can see it’s important to you, so it’s a great opportunity to contribute!

Hey @mattbaker, fair enough, actually that sounds like a great idea, flooded with work and parenting but this could be a good off the weekend project, no promises but if I make it work before some else does with my limited time, I'll be back. Thanks.

**Comment by scorphus** [comment]
@Esteban-Rocha hopefully you may have some time to kindly validate the changes I proposed? Just earlier last week, I was again faced with the lack of this feature in Zed. I hope not anymore.
